src/server/pkg/storage/fileset/storage.go

```diff
@@ -287,6 +318,92 @@ func (s *Storage) WalkFileSet(ctx context.Context, prefix string, f func(string)
 	})
 }
 
+// SetTTL sets the time-to-live for the prefix. If the prefix contains multiple filesets
+// the ttl will be set on all of them and the soonest expiration time will be returned.
+// if no fileset is found SetTTL returns ErrNoFileSetFound
+func (s *Storage) SetTTL(ctx context.Context, prefix string, ttl time.Duration) (time.Time, error) {
+	// TODO: after we move the semantic paths to postrges the whole prefix can be set atomically.
+	var expiresAt *time.Time
+	if err := s.objC.Walk(ctx, applyPrefix(prefix), func(p string) error {
+		t, err := s.chunks.RenewReference(ctx, p, ttl)
+		if err != nil {
+			return err
+		}
+		if expiresAt == nil || t.Before(*expiresAt) {
+			expiresAt = &t
+		}
+		return nil
+	}); err != nil {
+		return time.Time{}, err
+	}
+	if expiresAt == nil {
+		return time.Time{}, ErrNoFileSetFound
+	}
+	return *expiresAt, nil
+}
+
+// RenewDuring ensures that none of the filesets under prefix expire by repeatedly calling SetTTL in a separate goroutine
+// until cb returns.
+// RenewDuring attempts to delete the contents of prefix after cb has returned
+func (s *Storage) RenewDuring(ctx context.Context, prefix string, cb func(context.Context) error) error {
```

Comment:
I don't think this model is going to work for us. I think we need to be able to associate one `expiresAt` with the prefix rather than one for each individual path under that prefix. When we move to postgres, we will be cleaning up the semantic path, so I can imagine a scenario where a path under the prefix expires and is deleted, but we don't catch that it was deleted during the renewal since the renewal just looks at the existing paths. Or, we could just be explicit about which paths to renew, but I kind of like the model where we are renewing on a prefix.

Reply:
This has changed to take a list of paths, each of which will be passed to `SetTTL`